tests/sentry/nodestore/bigtable/backend/tests.py

```diff
@@ -51,22 +54,53 @@ def mutate_rows(self, rows):
             # commits not implemented, changes are applied immediately
             return [Status(code=0) for row in rows]
 
-    @memoize
-    def connection(self):
-        return MockedBigtableNodeStorage.Connection()
+    def _get_table(self, admin: bool = False):
+        try:
+            table = self.__table
+        except AttributeError:
+            table = self.__table = MockedBigtableKVStorage.Table()
 
-    def bootstrap(self):
+        return table
+
+    def bootstrap(self, automatic_expiry):
         pass
 
 
+class MockedBigtableNodeStorage(BigtableNodeStorage):
+    store_class = MockedBigtableKVStorage
+
+
+@contextmanager
+def get_temporary_bigtable_nodestorage() -> BigtableNodeStorage:
+    if "BIGTABLE_EMULATOR_HOST" not in os.environ:
+        pytest.skip(
+            "Bigtable is not available, set BIGTABLE_EMULATOR_HOST enironment variable to enable"
+        )
+
+    # The bigtable emulator requires _something_ to be passed as credentials,
+    # even if they're totally bogus ones.
+    ns = BigtableNodeStorage(
+        project="test",
+        credentials=Credentials.from_authorized_user_info(
+            {key: "invalid" for key in ["client_id", "refresh_token", "client_secret"]}
+        ),
+    )
+
+    ns.bootstrap()
+
+    try:
+        yield ns
+    finally:
+        ns.store._get_table(admin=True).delete()
+
+
 @pytest.fixture(params=[MockedBigtableNodeStorage, BigtableNodeStorage])
 def ns(request):
     if request.param is BigtableNodeStorage:
-        pytest.skip("Bigtable is not available in CI")
-
-    ns = request.param(project="test")
-    ns.bootstrap()
-    return ns
+        with get_temporary_bigtable_nodestorage() as ns:
+            yield ns
+    else:
+        yield MockedBigtableNodeStorage(project="test")
```

Comment:
I'm not really sure the mock backend is providing anything useful enough to justify its inclusion once/if the emulator is included in CI. The emulator is easy to run, better reflects real behavior (e.g. the cell timestamp), and is going to be a more representative test as it actually exercises the real client paths (e.g. highlighting the deprecation of `Table.row`.)

I think if you are modifying the storage behavior, you should be testing with something as close to the real storage as is practical. I do think that using actual Cloud Bigtable is impractical (due to the level of required permissions), but the emulator isn't and is going to be a lot closer to the real implementation than our 50 lines of mocks.

Reply:
agreed